File: F#/OrderTaking/Domain.fs
namespace OrderTaking.Domain

// 製品コード関連
/// TODO: 制約: 先頭が "W" + 数字4桁
type WidgetCode = WidgetCode of string
/// TODO: 制約: 先頭が "W" + 数字4桁
type GizmoCode = GizmoCode of string

type ProductCode =
    | Widget of WidgetCode
    | Gizmo of GizmoCode

// 注文数量関連
type UnitQuantity = UnitQuantity of int
type KilogramQuantity = KilogramQuantity of decimal

type OrderQuantity =
    | Unit of UnitQuantity
    | Kilos of KilogramQuantity

// 注文関連の識別子
type OrderId = Undefined
type OrderLineId = Undefined
type CustomerId = Undefined

// 注文関連
type CustomerInfo = Undefined
type ShippingAddress = Undefined
type BillingAddress = Undefined
type Price = Undefined
type BillingAmount = Undefined

/// Order集約のルート
type Order =
    { Id: OrderId
      CustomerId: CustomerId
      ShippingAddress: ShippingAddress
      BillingAddress: BillingAddress
      OrderLines: OrderLine list
      AmountToBill: BillingAmount }

and OrderLine =
    { id: OrderLineId
      OrderId: OrderId
      ProductCode: ProductCode
      OrderQuantity: OrderQuantity
      Price: Price }

// ワークフローの構成要素
type UnvalidatedOrder =
    { OrderId: string
      CustomerInfo: string
      ShippingAddress: string
      BillingAddress: string
      OrderLines: UnvalidatedOrderLine list }

and UnvalidatedOrderLine =
    { ProductCode: string
      OrderQuantity: string }

type PlaceOrderEvents =
    { AcknowledgmentSent: bool
      OrderPlaced: Order
      BillableOrderPlaced: BillableOrderPlaced }

and BillableOrderPlaced =
    { OrderId: OrderId
      BillingAddress: BillingAddress
      AmountToBill: BillingAmount }

type PlaceOrderError = ValidationError of ValidationError list

and ValidationError =
    { FieldName: string
      ErrorDescription: string }

// 注文確定プロセス
type PlaceOrder = UnvalidatedOrder -> Result<PlaceOrderEvents, PlaceOrderError>

File: Scala/ordertaking/src/main/scala/Domain.scala
// TODO: コンストラクタを隠蔽したり、apply のすり抜けを気にしないといけない感じになってくるなら、opaque の方が良さそう

// 製品コード関連
/// TODO: 先頭が "W" + 数字4桁
case class WidgetCode(value: String)

/// TODO: 先頭が "G" + 数字3桁
case class GizmoCode(value: String)

enum ProductCode:
  case Widget(code: WidgetCode)
  case Gizmo(code: GizmoCode)

// 注文数量関連
case class UnitQuantity(value: Int)

case class KilogramQuantity(value: BigDecimal)

enum OrderQuantity:
  case Unit(value: UnitQuantity)
  case Kilos(value: KilogramQuantity)

// 注文関連の識別子
// TODO: Any は未確定
case class OrderId(value: Any)

case class OrderLineId(value: Any)

case class CustomerId(value: Any)

// 注文関連
// TODO: Any は未確定
case class CustomerInfo(value: Any)

case class ShippingAddress(value: Any)

case class BillingAddress(value: Any)

case class Price(value: Any)

case class BillingAmount(value: Any)

// Order集約
case class OrderLine(
                      id: OrderLineId,
                      orderId: OrderId,
                      productCode: ProductCode,
                      orderQuantity: OrderQuantity,
                      price: Price,
                    )

case class Order(
                  id: OrderId,
                  customerInfo: CustomerInfo,
                  shippingAddress: ShippingAddress,
                  billingAddress: BillingAddress,
                  orderLines: Seq[OrderLine],
                  amountToBill: BillingAmount,
                )

// ワークフローの構成要素
case class UnvalidatedOrderLine(
                                 id: String,
                                 orderId: String,
                                 productCode: String,
                                 orderQuantity: String,
                               )

case class UnvalidatedOrder(
                             id: String,
                             customerInfo: String,
                             shippingAddress: String,
                             billingAddress: String,
                             orderLines: Seq[UnvalidatedOrderLine],
                           )

case class BillableOrderPlaced(
                                orderId: OrderId,
                                billingAddress: BillingAddress,
                                amountToBill: BillingAmount,
                              )

case class PlaceOrderEvents(
                             acknowledgementSent: Boolean,
                             orderPlaced: Order,
                             billableOrderPlaced: BillableOrderPlaced,
                           )

case class ValidationErrors(fieldName: String, errorDescription: String)

case class PlaceOrderError(errors: Seq[ValidationErrors])

// 注文確定プロセス
type PlaceOrder = UnvalidatedOrder => Either[PlaceOrderError, PlaceOrderEvents]
